# Retail Demand & Market Intelligence Analytics

# Phase 1 – Data Understanding

## Business Objective

The objective of this project is to analyze historical retail transaction data to understand customer purchasing behavior, identify seasonal demand patterns, and ultimately compare those patterns with retail-sector stock performance.

The insights from this project will support business decisions related to inventory planning, customer segmentation, demand forecasting, and strategic planning.

## Import Required Libraries

In [40]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)

## Load the Dataset

In [41]:
retail = pd.read_csv(
    "../data/raw/online_retail_II.csv",
    encoding="ISO-8859-1"
)
retail.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


## Performing initial analysis
# observation



In [42]:
retail.shape

(1067371, 8)

Observation: The dataset contains 1,067,371 transaction records across eight variables. This indicates that the dataset captures a large volume of retail activity over multiple years, making it suitable for analyzing customer purchasing behavior, seasonal demand patterns, and long-term business trends.

In [43]:
retail.describe(include="all")

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
count,1067371,1067371,1062989,1.067371e+06,1067371,1.067371e+06,824364.000000,1067371
unique,53628,5305,5698,NaN,47635,NaN,NaN,43
top,537434,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,2010-12-06 16:57:00,NaN,NaN,United Kingdom
freq,1350,5829,5918,NaN,1350,NaN,NaN,981330
mean,NaN,NaN,NaN,9.938898e+00,NaN,4.649388e+00,15324.638504,NaN
std,NaN,NaN,NaN,1.727058e+02,NaN,1.235531e+02,1697.464450,NaN
min,NaN,NaN,NaN,-8.099500e+04,NaN,-5.359436e+04,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000e+00,NaN,1.250000e+00,13975.000000,NaN
50%,NaN,NaN,NaN,3.000000e+00,NaN,2.100000e+00,15255.000000,NaN
75%,NaN,NaN,NaN,1.000000e+01,NaN,4.150000e+00,16797.000000,NaN


In [44]:
retail.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 65.1 MB


## Data Type Summary
**Observation:**
The dataset contains a mix of categorical (object), integer, and floating-point variables. Most columns are stored as object data types because they represent identifiers, textual information, or dates. Only Quantity is stored as an integer, while Price and Customer ID are stored as floating-point numbers.

**Business Insight:**
Before performing detailed analysis, some columns require data type conversion. For example, InvoiceDate should be converted to a datetime format to enable time-series analysis, while Customer ID should be treated as an identifier rather than a numerical variable.

## No of Duplicates

In [45]:
retail.duplicated().sum()

np.int64(34335)

In [46]:
(retail.duplicated().sum()/len(retail))*100

np.float64(3.2167821685243467)

## No of Null Values

In [47]:
retail.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

In [48]:
(retail.isnull().sum()/len(retail))*100

Invoice         0.000000
StockCode       0.000000
Description     0.410541
Quantity        0.000000
InvoiceDate     0.000000
Price           0.000000
Customer ID    22.766873
Country         0.000000
dtype: float64

## No of Unique values in each column

In [49]:
retail.nunique()

Invoice        53628
StockCode       5305
Description     5698
Quantity        1057
InvoiceDate    47635
Price           2807
Customer ID     5942
Country           43
dtype: int64

In [50]:
retail["Country"].value_counts()

Country
United Kingdom    981330
EIRE               17866
Germany            17624
France             14330
Netherlands         5140
                   ...  
West Indies           54
Bermuda               34
Nigeria               32
Czech Republic        30
Saudi Arabia          10
Name: count, Length: 43, dtype: int64

In [51]:
retail.groupby("StockCode")["Quantity"].sum().sort_values(ascending=False).head(10)

StockCode
84077     108545
85123A     96066
85099B     95739
21212      95450
84879      80705
22197      79363
17003      70700
21977      56575
84991      54366
22492      45384
Name: Quantity, dtype: int64

## Time Period

In [52]:
retail["InvoiceDate"]=pd.to_datetime(retail["InvoiceDate"])
retail["InvoiceDate"].min(), retail["InvoiceDate"].max()

(Timestamp('2009-12-01 07:45:00'), Timestamp('2011-12-09 12:50:00'))

## Cancelled Invoices

In [53]:
retail["Invoice"].str.startswith("C").sum()

np.int64(19494)

## Revenue

In [54]:
retail["Revenue"] = retail["Quantity"] * retail["Price"]
retail["Revenue"]

0           83.40
1           81.00
2           81.00
3          100.80
4           30.00
            ...  
1067366     12.60
1067367     16.60
1067368     16.60
1067369     14.85
1067370     18.00
Name: Revenue, Length: 1067371, dtype: float64

In [55]:
print(retail.columns.tolist())

['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country', 'Revenue']


# Analyzing the missing Customer IDs are random or concentrated somewhere specific (a country, a date range, an invoice type)

### i) Country distribution for rows with missing Customer IDs

In [56]:
retail[retail["Customer ID"].isna()]["Country"].value_counts()

Country
United Kingdom          240029
EIRE                      1671
Hong Kong                  364
Unspecified                232
France                     128
Switzerland                125
Portugal                   116
United Arab Emirates       114
Bahrain                     67
Israel                      47
RSA                         46
Bermuda                     34
Sweden                      19
Lebanon                     13
Nigeria                      2
Name: count, dtype: int64

### ii) Number of cancelled invoices among rows with missing Customer IDs and Percentage of missing Customer IDs that are cancelled invoices

In [57]:
retail[retail["Customer ID"].isna()]["Invoice"].str.startswith("C").sum()

np.int64(750)

In [58]:
(
    retail[retail["Customer ID"].isna()]["Invoice"].str.startswith("C").mean()
) * 100

np.float64(0.30863308464365224)

### iii) Monthly distribution of rows with missing Customer IDs

In [59]:
(
    retail[retail["Customer ID"].isna()]
    .groupby(retail["InvoiceDate"].dt.to_period("M"))
    .size()
)

InvoiceDate
2009-12    13468
2010-01     9116
2010-02     5482
2010-03     8397
2010-04     6224
           ...  
2011-08     7622
2011-09     9404
2011-10    10047
2011-11    19113
2011-12     7864
Freq: M, Length: 25, dtype: int64

# Display StockCodes that contain alphabetic characters


In [60]:
retail[
    retail["StockCode"].str.contains("[A-Za-z]", regex=True, na=False)
]["StockCode"].value_counts()

StockCode
85123A    5829
85099B    4216
POST      2122
82494L    2108
85099C    1952
          ... 
90184c       1
72802c       1
85049c       1
84971l       1
85034b       1
Name: count, Length: 1715, dtype: int64

In [74]:
retail[retail["StockCode"].str.fullmatch(r"[A-Za-z]+", na=False)]["StockCode"].value_counts()

StockCode
POST         2122
DOT          1446
M            1421
D             177
S             104
ADJUST         67
AMAZONFEE      43
DCGSSGIRL      25
DCGSSBOY       23
PADS           19
CRUK           16
B               6
m               5
GIFT            1
DCGSLBOY        1
DCGSLGIRL       1
Name: count, dtype: int64

**Observation:** The investigation identified several StockCodes consisting entirely of alphabetic characters that represent administrative or operational transactions rather than physical retail products. Examples include POST (Postage), M (Manual Adjustment), D (Discount), B (Bad Debt Adjustment), and AMAZONFEE. These entries should be treated separately from product sales during the Data Cleaning phase. A few alphabetic StockCodes (e.g., DCGSSGIRL, DCGSSBOY, PADS, and GIFT) may still represent genuine products or promotional items and should be verified before any exclusions are made.

In [75]:
retail[retail["Quantity"].isin([retail["Quantity"].max(), retail["Quantity"].min()])]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
1065882,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,United Kingdom,168469.6
1065883,C581484,23843,"PAPER CRAFT , LITTLE BIRDIE",-80995,2011-12-09 09:27:00,2.08,16446.0,United Kingdom,-168469.6


**Observation:** The largest positive and negative Quantity values (±80,995) belong to the same product, customer, and price, with the negative transaction recorded as a cancelled invoice shortly after the original purchase. This indicates that the extreme values represent a legitimate business transaction and its corresponding cancellation rather than a data entry error. The handling of cancelled transactions will be addressed during the Data Cleaning phase.

In [76]:
retail[retail["Price"] < 0]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom,-53594.36
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom,-44031.79
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom,-38925.87
825444,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom,-11062.06
825445,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom,-11062.06


**Observation:** All rows with negative prices correspond to the administrative StockCode 'B' with the description 'Adjust bad debt'. These are accounting adjustments rather than actual product sales. In addition, all such records have missing Customer IDs, indicating that they cannot be linked to individual customers. These entries will require special handling during the Data Cleaning phase and should not be treated as normal retail transactions.

In [77]:
retail[retail["StockCode"].str.fullmatch(r"[A-Za-z]+", na=False)]["StockCode"].value_counts()

StockCode
POST         2122
DOT          1446
M            1421
D             177
S             104
ADJUST         67
AMAZONFEE      43
DCGSSGIRL      25
DCGSSBOY       23
PADS           19
CRUK           16
B               6
m               5
GIFT            1
DCGSLBOY        1
DCGSLGIRL       1
Name: count, dtype: int64

**Observation:** The investigation identified several StockCodes consisting entirely of alphabetic characters that represent administrative or operational transactions rather than physical retail products. Examples include POST (Postage), M (Manual Adjustment), D (Discount), B (Bad Debt Adjustment), and AMAZONFEE. These entries should be treated separately from product sales during the Data Cleaning phase. A few alphabetic StockCodes (e.g., DCGSSGIRL, DCGSSBOY, PADS, and GIFT) may still represent genuine products or promotional items and should be verified before any exclusions are made.

In [78]:
(retail.groupby("StockCode")["Description"].nunique().sort_values(ascending=False).head(20))

StockCode
20713     9
22734     7
23084     7
22423     7
21181     7
85175     6
47566B    6
22719     6
21830     6
85123A    5
22501     5
85172     5
72807A    5
22740     5
22502     5
23343     5
23131     5
21843     5
22467     4
22785     4
Name: Description, dtype: int64

**Observation:** Some StockCodes are associated with multiple Description values due to data entry mistakes and manual correction records rather than representing different products.

In [79]:
retail[retail.duplicated(keep=False)].sort_values(["Invoice", "StockCode"]).head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom,1.95
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom,1.95
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,3.75
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,3.75
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,3.75
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,3.75
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,3.75
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,3.75
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom,3.75
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom,5.10


**Observation:** Inspection of duplicate records confirms that all duplicated rows are identical across every column, including the invoice number, product, quantity, price, timestamp, customer, and country. These duplicates are therefore likely caused by data entry or system duplication rather than genuine repeated purchases. Retaining them would result in double-counting of sales, revenue, and product demand. These records should be removed during the Data Cleaning phase.

In [80]:
(retail["Quantity"]==0).sum()

np.int64(0)

In [81]:
(retail["Price"]==0).sum()

np.int64(6202)

In [82]:
retail[retail["Price"] == 0].head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom,-0.0
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom,-0.0
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom,-0.0
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom,-0.0
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom,-0.0
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom,0.0
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,NaN,United Kingdom,-0.0
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,NaN,United Kingdom,-0.0
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom,0.0
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom,-0.0


In [83]:
retail[retail["Price"] == 0][["Description"]].value_counts().head(20)

Description                        
check                                  162
?                                       92
damages                                 84
damaged                                 81
found                                   28
missing                                 27
sold as set on dotcom                   20
Damaged                                 17
adjustment                              16
OWL DOORSTOP                            15
POLYESTER FILLER PAD 45x45cm            12
dotcom                                  12
amazon                                  11
POLYESTER FILLER PAD 40x40cm            10
IVORY KITCHEN SCALES                    10
FRENCH BLUE METAL DOOR SIGN 1           10
smashed                                  9
Found                                    9
PICNIC BASKET WICKER LARGE               9
AIRLINE BAG VINTAGE WORLD CHAMPION       9
Name: count, dtype: int64

In [84]:
retail[retail["Price"] == 0]["Customer ID"].isna().mean() * 100

np.float64(98.8552079974202)

In [85]:
retail[retail["Invoice"] == "537434"]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
516208,537434,20685,DOORMAT RED RETROSPOT,1,2010-12-06 16:57:00,14.43,NaN,United Kingdom,14.43
516209,537434,20699,MOUSEY LONG LEGS SOFT TOY,1,2010-12-06 16:57:00,5.06,NaN,United Kingdom,5.06
516210,537434,20713,JUMBO BAG OWLS,1,2010-12-06 16:57:00,4.21,NaN,United Kingdom,4.21
516211,537434,20719,WOODLAND CHARLOTTE BAG,3,2010-12-06 16:57:00,1.66,NaN,United Kingdom,4.98
516212,537434,20725,LUNCH BAG RED RETROSPOT,4,2010-12-06 16:57:00,4.21,NaN,United Kingdom,16.84
...,...,...,...,...,...,...,...,...,...
539401,537434,20652,BLUE POLKADOT LUGGAGE TAG,2,2010-12-06 16:57:00,2.51,NaN,United Kingdom,5.02
539402,537434,20658,RED RETROSPOT LUGGAGE TAG,1,2010-12-06 16:57:00,2.51,NaN,United Kingdom,2.51
539403,537434,20665,RED RETROSPOT PURSE,1,2010-12-06 16:57:00,5.91,NaN,United Kingdom,5.91
539404,537434,20682,RED RETROSPOT CHILDRENS UMBRELLA,1,2010-12-06 16:57:00,6.77,NaN,United Kingdom,6.77


**Observation:** Invoice 537434 contains 1,350 product line items, making it the largest invoice in the dataset. All records share the same invoice number, timestamp, and country, indicating a legitimate large wholesale transaction rather than duplicated or corrupted data. However, the invoice has a missing Customer ID, reinforcing the earlier finding that valid retail transactions may still lack customer identifiers.

## Descriptive statistics on numeric columns(Quantity ,Price)

In [86]:
retail.Quantity.describe()

count    1.067371e+06
mean     9.938898e+00
std      1.727058e+02
min     -8.099500e+04
25%      1.000000e+00
50%      3.000000e+00
75%      1.000000e+01
max      8.099500e+04
Name: Quantity, dtype: float64

In [87]:
retail.Price.describe()

count    1.067371e+06
mean     4.649388e+00
std      1.235531e+02
min     -5.359436e+04
25%      1.250000e+00
50%      2.100000e+00
75%      4.150000e+00
max      3.897000e+04
Name: Price, dtype: float64

In [90]:
retail.describe(include="all")

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
count,1067371,1067371,1062989,1.067371e+06,1067371,1.067371e+06,824364.000000,1067371,1.067371e+06
unique,53628,5305,5698,NaN,NaN,NaN,NaN,43,NaN
top,537434,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,NaN,NaN,NaN,United Kingdom,NaN
freq,1350,5829,5918,NaN,NaN,NaN,NaN,981330,NaN
mean,NaN,NaN,NaN,9.938898e+00,2011-01-02 21:13:55.394029,4.649388e+00,15324.638504,NaN,1.806987e+01
min,NaN,NaN,NaN,-8.099500e+04,2009-12-01 07:45:00,-5.359436e+04,12346.000000,NaN,-1.684696e+05
25%,NaN,NaN,NaN,1.000000e+00,2010-07-09 09:46:00,1.250000e+00,13975.000000,NaN,3.750000e+00
50%,NaN,NaN,NaN,3.000000e+00,2010-12-07 15:28:00,2.100000e+00,15255.000000,NaN,9.900000e+00
75%,NaN,NaN,NaN,1.000000e+01,2011-07-22 10:23:00,4.150000e+00,16797.000000,NaN,1.770000e+01
max,NaN,NaN,NaN,8.099500e+04,2011-12-09 12:50:00,3.897000e+04,18287.000000,NaN,1.684696e+05


In [91]:
retail.info()


<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 9 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  str           
 1   StockCode    1067371 non-null  str           
 2   Description  1062989 non-null  str           
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[us]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  str           
 8   Revenue      1067371 non-null  float64       
dtypes: datetime64[us](1), float64(3), int64(1), str(4)
memory usage: 73.3 MB
